File: ScanProject/JavaAnalysis/SonarJava/src/main/java/Medium/831-matchsticks-to-square.java
package main.java.Medium;

/**https://leetcode.com/problems/matchsticks-to-square/ */
//You are given an integer array matchsticks where matchsticks[i] is the length of the ith matchstick. You want to use all the matchsticks to make one square. You should not break any stick, but you can link them up, and each matchstick must be used exactly one time.
//Return true if you can make this square and false otherwise.
// 
//Example 1:
//Input: matchsticks = [1,1,2,2,2]
//Output: true
//Explanation: You can form a square with length 2, one side of the square came two sticks with length 1.
//Example 2:
//Input: matchsticks = [3,3,3,3,4]
//Output: false
//Explanation: You cannot find a way to form a square with all the matchsticks.
// 
//Constraints:
//	1 <= matchsticks.length <= 15
//	1 <= matchsticks[i] <= 108
class Solution495 {
    public boolean makesquare(int[] matchsticks) {
        int sum = 0;
        for (int i : matchsticks) {
            sum += i;
        }
        if (sum % 4 != 0) {
            return false;
        }

        int[] sides = new int[4];
        return dfs(matchsticks, 0, sides, sum / 4);
    }

    private boolean dfs(int[] matchsticks, int index, int[] sides, int target) {
        if (index == matchsticks.length) {
            return sides[0] == sides[1] && sides[1] == sides[2] && sides[2] == sides[3];
        }
        for (int i = 0; i < 4; i++) {
            if (sides[i] + matchsticks[index] > target) {
                continue;
            }
            sides[i] += matchsticks[index];
            if (dfs(matchsticks, index + 1, sides, target)) {
                return true;
            }
            sides[i] -= matchsticks[index];
        }
        return false;
    }
}

File: ScanProject/JSAnalysis/SonarJS/Medium/831-matchsticks-to-square.js
/**https://leetcode.com/problems/matchsticks-to-square/ */
//You are given an integer array matchsticks where matchsticks[i] is the length of the ith matchstick. You want to use all the matchsticks to make one square. You should not break any stick, but you can link them up, and each matchstick must be used exactly one time.
//Return true if you can make this square and false otherwise.


/**
 * @param {number[]} matchsticks
 * @return {boolean}
 */
var makesquare = function(matchsticks) {
    let n = matchsticks.length;
    if(n < 4){
        return false;
    }
    let sum = 0;
    for(let i = 0; i < n; i++){
        sum += matchsticks[i];
    }
    if(sum % 4 != 0){
        return false;
    }
    let target = sum / 4;
    matchsticks.sort((a, b) => b - a);
    let sides = new Array(4).fill(0);
    return dfs(matchsticks, sides, 0, target);
    
    
};
function dfs(matchsticks, sides, index, target){
    if(index == matchsticks.length){
        return sides[0] == sides[1] && sides[1] == sides[2] && sides[2] == sides[3];
    }
    for(let i = 0; i < 4; i++){
        if(sides[i] + matchsticks[index] <= target){
            sides[i] += matchsticks[index];
            if(dfs(matchsticks, sides, index + 1, target)){
                return true;
            }
            sides[i] -= matchsticks[index];
        }
    }
    return false;
}

    